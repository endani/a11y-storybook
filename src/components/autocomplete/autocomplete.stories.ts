import type { Meta, StoryObj } from "@storybook/react"
import { Autocomplete } from "./autocomplete"
import { fn } from "@storybook/test"

const mockResult = [
  "Apple",
  "Banana",
  "Cherry",
  "Date",
  "Elderberry",
  "Fig",
  "Grape",
  "Honeydew",
]

const meta = {
  title: "Example/AutoComplete",
  component: Autocomplete,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {},
  args: {
    placeholder: "Search for a fruit",
    onChange: fn(),
    onResultSelect: fn(),
  },
} satisfies Meta<typeof Autocomplete>

export default meta
type Story = StoryObj<typeof meta>

export const Primary: Story = {
  args: {},
}

export const WithError: Story = {
  args: {
    hasError: true,
  },
}

export const WithLoading: Story = {
  args: {
    isLoading: true,
  },
}

export const WithSelectedValue: Story = {
  args: {
    selectedValue: "Apple",
  },
}

export const WithResults: Story = {
  args: {
    options: mockResult,
  },
}
