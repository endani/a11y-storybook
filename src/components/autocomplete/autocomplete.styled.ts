import styled from "styled-components"

export const AutocompleteWrapper = styled.form`
  align-items: flex-end;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  gap: 10px;
  margin-bottom: 20px;
  width: 100%;
`
