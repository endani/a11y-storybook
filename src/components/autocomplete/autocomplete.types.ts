export interface AutocompleteProps {
  autoFocus?: boolean
  hasError?: boolean
  isLoading?: boolean
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void
  onResultSelect: (option: string) => void
  options?: string[]
  placeholder: string
  searchTerm?: string | null
  selectedValue?: string | null
}
