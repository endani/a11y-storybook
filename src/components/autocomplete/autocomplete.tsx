import { useRef } from "react"
import { Input, ErrorMessage, List } from "../index"

import { AutocompleteWrapper } from "./autocomplete.styled"
import { AutocompleteProps } from "./autocomplete.types"

const Autocomplete = ({
  autoFocus = true,
  hasError = false,
  isLoading,
  onChange,
  onResultSelect,
  options,
  placeholder,
  searchTerm = "",
  selectedValue,
}: AutocompleteProps) => {
  const inputRef = useRef<HTMLInputElement>(null)

  const handleOnResultSelect = (option: string) => {
    inputRef.current!.value = option
    onResultSelect(option)
  }

  return (
    <div data-qa="autocomplete">
      <AutocompleteWrapper>
        <Input
          aria-autocomplete="list"
          aria-controls="autocomplete-list"
          aria-expanded={!!options?.length}
          autoComplete="off"
          autoFocus={autoFocus}
          inputRef={inputRef}
          onChange={onChange}
          placeholder={placeholder}
          role="combobox"
          tabIndex={0}
          type="search"
        />
        {!isLoading && !!options?.length && (
          <List
            id="autocomplete-list"
            options={options}
            onSelect={handleOnResultSelect}
            searchTerm={searchTerm}
            maxResults={5}
          />
        )}
      </AutocompleteWrapper>

      {isLoading && <ErrorMessage>Loading...</ErrorMessage>}
      {!isLoading && hasError && (
        <>
          <ErrorMessage>Ops! Something went wrong. Try again</ErrorMessage>
        </>
      )}
      {!isLoading && !hasError && searchTerm && options?.length === 0 && (
        <ErrorMessage>
          No results found for <b>{searchTerm}</b>. Will ask our chef to add it
          to the menu! 👩‍🍳
        </ErrorMessage>
      )}
      {selectedValue && (
        <ErrorMessage>
          You have selected <b>{selectedValue}</b> 🎉
        </ErrorMessage>
      )}
    </div>
  )
}

export { Autocomplete }
