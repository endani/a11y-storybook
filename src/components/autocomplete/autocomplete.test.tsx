import { render, screen } from "@testing-library/react"

import { Autocomplete } from "./autocomplete"
import { AutocompleteProps } from "./autocomplete.types"

const mockOnChange = jest.fn()

const renderAutocomplete = ({ placeholder, ...props }: AutocompleteProps) =>
  render(
    <Autocomplete
      {...props}
      placeholder={placeholder}
      onChange={mockOnChange}
    />
  )

describe("Autocomplete", () => {
  it("should render a Autocomplete correctly", () => {
    const { container } = renderAutocomplete({
      placeholder: "Awesome Autocomplete",
    } as AutocompleteProps)

    expect(container).toMatchSnapshot()
    expect(screen.getByTestId("autocomplete")).toBeInTheDocument()
  })

  it("should not show the error message by default", () => {
    renderAutocomplete({
      placeholder: "Awesome Autocomplete",
    } as AutocompleteProps)

    expect(screen.queryByTestId("error-message")).not.toBeInTheDocument()
  })

  it("should show the error message when hasError is true", async () => {
    renderAutocomplete({
      placeholder: "Awesome Autocomplete",
      hasError: true,
    } as AutocompleteProps)

    expect(screen.queryByTestId("error-message")).toBeInTheDocument()
    expect(
      screen.getByText("Ops! Something went wrong. Try again")
    ).toBeInTheDocument()
  })

  it("should show the loading message when isLoading is true", async () => {
    renderAutocomplete({
      placeholder: "Awesome Autocomplete",
      isLoading: true,
    } as AutocompleteProps)

    expect(screen.queryByTestId("error-message")).toBeInTheDocument()
    expect(screen.getByText("Loading...")).toBeInTheDocument()
  })
})
