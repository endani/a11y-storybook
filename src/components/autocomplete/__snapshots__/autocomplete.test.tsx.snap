// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete should render a Autocomplete correctly 1`] = `
<div>
  <div
    data-qa="autocomplete"
  >
    <form
      class="sc-gsFSjX dIyWkF"
    >
      <input
        aria-autocomplete="list"
        aria-controls="autocomplete-list"
        aria-expanded="false"
        autocomplete="off"
        class="sc-eqUzNf kvHAoH"
        data-qa="input"
        placeholder="Awesome Autocomplete"
        role="combobox"
        tabindex="0"
        type="search"
      />
    </form>
  </div>
</div>
`;
